gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageProvider.java

```diff
@@ -0,0 +1,196 @@
+
+  /** Reverse map for storage reference to cache keys. */
+  final Map<Storage, StorageProviderCacheKey> storageToCacheKeyMap = new ConcurrentHashMap<>();
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting
+  StorageProviderCacheKey computeCacheKey(
+      Credentials credentials,
+      GoogleCloudStorageOptions storageOptions,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn) {
+    return StorageProviderCacheKey.builder()
+        .setCredentials(credentials)
+        .setHttpHeaders(storageOptions.getHttpRequestHeaders())
+        .setIsDirectPathPreferred(storageOptions.isDirectPathPreferred())
+        .setIsDownScopingEnabled(downscopedAccessTokenFn != null)
+        .setIsTracingEnabled(storageOptions.isTraceLogEnabled())
+        .setWriteChannelOptions(storageOptions.getWriteChannelOptions())
+        .setProjectId(storageOptions.getProjectId())
+        .build();
+  }
+
+  /** Determines if the storage instance can be served from the cache. */
+  private static boolean canCache(
+      GoogleCloudStorageOptions options,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService) {
+
+    return options.isStorageClientCachingExperimentEnabled()
+        // These values are currently passed while creating the storage client, but they are
+        // currently not configurable by the FS options so skipping caching.
+        && pCUExecutorService == null
+        && (interceptors == null || interceptors.isEmpty());
+  }
+
+  Storage getStorage(
+      Credentials credentials,
+      GoogleCloudStorageOptions storageOptions,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn)
+      throws IOException {
+    if (canCache(storageOptions, interceptors, pCUExecutorService)) {
+      StorageProviderCacheKey key =
+          computeCacheKey(credentials, storageOptions, downscopedAccessTokenFn);
+      Storage storage = cache.getIfPresent(key);
+      if (storage == null) {
+        storage = createStorage(credentials, storageOptions, null, null, downscopedAccessTokenFn);
+        cache.put(key, storage);
+        storageToCacheKeyMap.put(storage, key);
+        logger.atFinest().log(
+            "Cache miss, created new storage client. Cache hit count : %d, Cache hit rate : %d",
+            cache.stats().hitCount(), cache.stats().hitRate());
+      } else {
+        logger.atFinest().log(
+            "Cache hit, reusing the storage client. Cache hit count : %d, Cache hit rate : %d",
+            cache.stats().hitCount(), cache.stats().hitRate());
+      }
+      // Increment the reference count of the storage object.
+      storageClientToReferenceMap.put(
+          storage, storageClientToReferenceMap.getOrDefault(storage, 0) + 1);
+      return storage;
+    }
+    return createStorage(
```

Comment:
we can return this in if check first, this way we can avoid the large if block:
```
if (!canCache(...)) {
  return ...;
}
// rest of code.
```

Reply:
Done